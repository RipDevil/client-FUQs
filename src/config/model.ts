import { createStore, createEvent } from 'effector';
import defaultConfig from './defaultConfig';

export interface ConfigType {
  server: string;
  api: {
    fuq?: {
      get?: string;
      put?: string;
      post?: string;
      delete?: string;
    };
  };
}

const $config = createStore<ConfigType>(defaultConfig);

const configUpdate = createEvent<ConfigType>('update config');

$config.on(
  configUpdate,
  (state, payload): ConfigType => ({ ...state, ...payload })
);

export { configUpdate, $config };
