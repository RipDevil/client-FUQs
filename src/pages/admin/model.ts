import { FuqType } from 'pages/fuq/model';

export interface UserType {
  login: string;
  fuqs: FuqType[];
  password: string;
  deleted: boolean;
} // password is presented? lol?
