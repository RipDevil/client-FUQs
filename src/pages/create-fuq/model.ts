export interface ReducedFuqType {
  title: string;
  text: string;
}
