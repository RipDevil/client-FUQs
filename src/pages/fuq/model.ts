export interface FuqType {
  _id: string;
  likes: number;
  _lastEditor: string;
  pending: boolean;
  title: string;
  text: string;
  crdate: string;
};
