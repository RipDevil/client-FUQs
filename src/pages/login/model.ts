import { createStore, createEvent } from 'effector';

export interface AuthType {
  token: string;
  refreshToken: string;
}

const defaultAuth = {
  token: '',
  refreshToken: '',
};

const $auth = createStore<AuthType>(defaultAuth);

const authUpdate = createEvent<AuthType>('update auth data (token and refresh token)');
const authReset = createEvent<any>('reset auth data (token and refresh token)');

$auth.on(authUpdate, (state, payload): AuthType => ({ ...state, ...payload })).reset(authReset);

export { authUpdate, authReset, $auth };
