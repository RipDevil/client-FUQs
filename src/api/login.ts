import { AxiosError } from 'axios';
import { useMutation } from 'react-query';

import { $config } from 'config/model';
import { call } from './lib';
import { AuthType } from 'pages/login/model';

interface CredentialsType {
  login: string;
  password: string;
}

/**
 * Perform login
 * @param credentials - { login, password }
 */
export const useLogin = () => {
  const {
    api: { auth },
    server,
  } = $config.getState();

  return useMutation<AuthType, AxiosError, CredentialsType, AuthType>('Login', (credentials) =>
    call(`${server}${auth?.login}`, 'POST', credentials),
  );
};
