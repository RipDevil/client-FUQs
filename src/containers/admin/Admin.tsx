import * as React from 'react';
import Layout from 'containers/layout/Layout';

export interface AdminProps {}

const Admin: React.FC<AdminProps> = () => {
  return (
    <Layout>
      <b>/badmin</b>
    </Layout>
  );
};

export default Admin;
