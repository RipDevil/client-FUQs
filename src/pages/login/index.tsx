import * as React from 'react';
import H from 'history';
import { useStore } from 'effector-react';
import { useHistory } from 'react-router-dom';
import { Form, Input, Button, Alert, Row, Col } from 'antd';
import { UserOutlined, LockOutlined } from '@ant-design/icons';

import { Layout } from 'templates';
import { Spinner } from 'components/common';
import { useLogin } from 'api';
import { login as loginRules, password as passRules } from './lib';

import { authUpdate, $auth } from './model';

type FormTypes = {
  login: string;
  password: string;
};

const Login: React.FC = () => {
  const loginInputRef = React.createRef<Input>();

  const history = useHistory<H.History>();
  const authStore = useStore($auth);
  const { mutate: login, isError, isSuccess, data, isLoading, error } = useLogin();

  React.useEffect(() => {
    if (authStore.token !== '' && authStore.refreshToken !== '') {
      history.push('/badmin');
    }
  });

  React.useEffect(() => {
    loginInputRef && loginInputRef.current?.focus();
  }, [loginInputRef]);

  React.useEffect(() => {
    if (!isError && isSuccess && data) {
      authUpdate(data);
    }
  }, [isSuccess, isError, data]);

  const onFormFinished = (credentials: FormTypes) => {
    login(credentials);
  };

  return (
    <Layout>
      <Col span="6">
        <Row justify="center" style={{ margin: '10px' }}>
          {isLoading && <Spinner text="In process" />}
          {isError && (
            <Alert
              message={`${error?.name}:` || 'Error: '}
              description={error?.message || 'Some error has occuered'}
              type="error"
              showIcon
              closable
            />
          )}
          {isSuccess && (
            <Alert message="Success" description="Successfully logged in" type="success" showIcon closable />
          )}
        </Row>

        <Form
          size="large"
          wrapperCol={{
            span: 24,
          }}
          onFinish={onFormFinished}
        >
          <Form.Item name="login" rules={loginRules}>
            <Input
              ref={loginInputRef}
              data-testid="login-input"
              maxLength={31}
              prefix={<UserOutlined className="site-form-item-icon" />}
              placeholder="Login"
            />
          </Form.Item>

          <Form.Item name="password" rules={passRules}>
            <Input.Password
              data-testid="password-input"
              maxLength={31}
              prefix={<LockOutlined className="site-form-item-icon" />}
              placeholder="Password"
            />
          </Form.Item>

          <Form.Item>
            <Button data-testid="button-login" type="primary" htmlType="submit" block>
              Login
            </Button>
          </Form.Item>
        </Form>
      </Col>
    </Layout>
  );
};

export default Login;
