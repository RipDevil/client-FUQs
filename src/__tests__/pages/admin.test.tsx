import { QueryClient, QueryClientProvider } from 'react-query';
import { Router, Route } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { render, waitFor, cleanup, fireEvent } from '@testing-library/react';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';
import { authUpdate, authReset } from 'pages/login/model';

import Admin from 'pages/admin';
import Login from 'pages/login';

const FAKE_CREDENTIALS = {
  token: 'FAKE_TOKEN',
  refreshToken: 'FAKE_REFRESH_TOKEN',
};

let queryClient: any;
let mockAxios: any;

beforeAll(() => {
  queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        refetchOnWindowFocus: false,
        retry: false,
      },
    },
  });
  mockAxios = new MockAdapter(axios);
});

afterEach(() => {
  mockAxios.reset();
  authReset();
  cleanup();
});

describe('Admin page tests', () => {
  it('Admin page should NOT be rendered because of bad credentials', async () => {
    const { container, getByTestId } = render(
      <QueryClientProvider client={queryClient}>
        <Router history={createMemoryHistory({ initialEntries: ['/badmin'] })}>
          <Route path={'/badmin'} component={Admin} />
          <Route path={'/login'} component={Login} />
        </Router>
      </QueryClientProvider>,
    );

    await waitFor(() => {
      const loginButton = getByTestId('button-login');
      return loginButton;
    });

    expect(container).toMatchSnapshot();
  });

  it('Admin page should be rendered with sub elements', async () => {
    authUpdate(FAKE_CREDENTIALS);

    mockAxios.onGet('/users').reply(200, []);

    const { container, getByText } = render(
      <QueryClientProvider client={queryClient}>
        <Router history={createMemoryHistory({ initialEntries: ['/badmin'] })}>
          <Route path={'/badmin'} component={Admin} />
        </Router>
      </QueryClientProvider>,
    );

    await waitFor(() => getByText('Actions'));

    expect(container).toMatchSnapshot();
  });

  it('User have to be redirected to the login page from admin page after logout', async () => {
    mockAxios.onPost('/auth/logout').reply(204, 'Success');
    authUpdate(FAKE_CREDENTIALS);
    const { getByTestId } = render(
      <QueryClientProvider client={queryClient}>
        <Router history={createMemoryHistory({ initialEntries: ['/badmin'] })}>
          <Route path={'/badmin'} component={Admin} />
          <Route path={'/login'} component={Login} />
        </Router>
      </QueryClientProvider>,
    );

    const logoutButton = getByTestId('logout-btn');
    fireEvent.click(logoutButton);

    await waitFor(() => {
      const loginButton = getByTestId('button-login');
      return loginButton;
    });
  });

  it('On a grip click the menu should change', async () => {
    authUpdate(FAKE_CREDENTIALS);
    const { getByTitle, getByTestId } = render(
      <QueryClientProvider client={queryClient}>
        <Router history={createMemoryHistory({ initialEntries: ['/badmin'] })}>
          <Route path={'/badmin'} component={Admin} />
        </Router>
      </QueryClientProvider>,
    );

    // @mna: default grip rn is the first one => 0
    expect(getByTitle('Users')).toBeTruthy();

    const itemButton = getByTestId('1');
    fireEvent.click(itemButton);

    await waitFor(() => {
      const fuqsGripTitle = getByTitle('FUQs');
      return fuqsGripTitle;
    });
  });
});
