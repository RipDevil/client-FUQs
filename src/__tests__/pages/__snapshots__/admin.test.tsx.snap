// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin page should be rendered 1`] = `
.c0 {
  padding-top: 40vh;
  text-align: center;
}

<div>
  <div
    class="ant-row ant-row-center ant-row-middle c0"
  >
    <b>
      /badmin
    </b>
  </div>
</div>
`;
